MaxText/layers/attentions.py

```diff
@@ -328,44 +328,132 @@ def attention_op_as_linen(
+    self.ragged_qkv_axis_names = ragged_qkv_axis_names
+    self.ragged_lengths_names = ragged_lengths_names
+    self.compute_axis_order = compute_axis_order
+    self.key_axis_order = key_axis_order
+    self.reshape_q = reshape_q
+    self.dropout_rate = dropout_rate
+    self.dtype = dtype
+    self.quant = quant
+    self.kv_quant = kv_quant
+    self.attention_type = attention_type
+    self.attn_logits_soft_cap = attn_logits_soft_cap
+    self.sliding_window_size = sliding_window_size
+    self.chunk_attn_window_size = chunk_attn_window_size
+    self.use_ragged_attention = use_ragged_attention
+    self.ragged_block_size = ragged_block_size
+
+    def maybe_create_nnx(einsum, *args):
+      if isinstance(einsum, nn.Module):
+        return nnx_wrappers.ToNNX(einsum, rngs=rngs).lazy_init(*args)
+      return einsum
+
+    # qk_product
+    if self.kv_quant:
+      # Dummy inputs for lazy initialization
+      b = 1
+      t_prefill = self.max_prefill_predict_length
+      t_ar = 1  # Autoregressive mode has a query length of 1
+      n = self.num_query_heads
+      n_kv = self.num_kv_heads
+      d = self.config.head_dim
+      g = n // n_kv
+      s_prefill = self.max_prefill_predict_length
+      s_ar = self.max_target_length
+
+      # Dummy query/key/value shapes as before...
+      dummy_query_prefill = jnp.zeros((b, t_prefill, n_kv, g, d), dtype=self.dtype)
+      dummy_key_prefill = jnp.zeros((b, s_prefill, n_kv, d), dtype=self.dtype)
+      dummy_query_ar = jnp.zeros((b, t_ar, n_kv, g, d), dtype=self.dtype)
+      dummy_key_ar = jnp.zeros((b, s_ar, n_kv, d), dtype=self.dtype)
+
+      dummy_attn_weights_prefill = jnp.zeros((b, n_kv, g, t_prefill, s_prefill), dtype=jnp.float32)
+      dummy_value_prefill = jnp.zeros((b, s_prefill, n_kv, d), dtype=self.dtype)
+      dummy_attn_weights_ar = jnp.zeros((b, n_kv, g, t_ar, s_ar), dtype=jnp.float32)
+      dummy_value_ar = jnp.zeros((b, s_ar, n_kv, d), dtype=self.dtype)
+
+      # Prefill AqtEinsum instances
+      self.AqtEinsum_0 = maybe_create_nnx(
+          self.kv_quant.einsum_fn_with_rhs_qtensor(),
+          "btkgd,bskd->bkgts", dummy_query_prefill, dummy_key_prefill
+      )
+      self.AqtEinsum_1 = maybe_create_nnx(
+          self.kv_quant.einsum_fn_with_rhs_qtensor_and_dequant(),
+          "bkgts,bskd->btkgd", dummy_attn_weights_prefill, dummy_value_prefill
+      )
+      # Autoregressive AqtEinsum instances
+      self.AqtEinsum_2 = maybe_create_nnx(
+          self.kv_quant.einsum_fn_with_rhs_qtensor(),
+          "btkgd,bskd->bkgts", dummy_query_ar, dummy_key_ar
+      )
+      self.AqtEinsum_3 = maybe_create_nnx(
+          self.kv_quant.einsum_fn_with_rhs_qtensor_and_dequant(),
+          "bkgts,bskd->btkgd", dummy_attn_weights_ar, dummy_value_ar
+      )
+    else:
+      self.AqtEinsum_0 = jnp.einsum
+      self.AqtEinsum_1 = jnp.einsum
+      self.AqtEinsum_2 = jnp.einsum
+      self.AqtEinsum_3 = jnp.einsum
```

Comment:
Why is this new logic needed?

Reply:
We need to explicitely define the `AqtEinsum_*` layers under `__init__` now, all this logic is mostly creating some dummy values for `lazy_init` to initialize the underlying `AqtEinsum` Linen Module.